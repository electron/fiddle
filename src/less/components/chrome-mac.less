.darwin {
  .chrome {
    text-align: center;
    background-color: @background-1;
    border-bottom: @border;
    -webkit-app-region: drag;
    height: 24px;
    font-size: 1.4rem;
    line-height: 15px;

    small,
    p,
    span {
      user-select: none;
    }
  }

  .big_sur_or_later {
    height: 30px !important;
  }
}
