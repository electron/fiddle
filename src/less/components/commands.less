@import "../variables.less";

header {
  box-shadow: 0 0 0 1px rgba(16, 22, 26, 0.2), 0 0 0 rgba(16, 22, 26, 0), 0 1px 1px rgba(16, 22, 26, 0.4);
  background-color: @background-3;
  -webkit-app-region: drag;
}

.disabled-menu-tooltip {
  display: flex;
  width: 100%;

  .bp3-popover-target {
    width: 100%;
  }
}

.commands.is-mac {
  padding:
    10px
    calc(100vw - (env(titlebar-area-x, 0) + env(titlebar-area-width, 0)))
    10px
    env(titlebar-area-x, 0);
}

.commands {
  display: flex;
  justify-content: space-between;
  height: 50px;
  padding:
    10px;
  box-sizing: border-box;
  font-family: @fonts-common;

  button,
  input,
  select {
    -webkit-app-region: no-drag;
    margin-bottom: 0;
    white-space: nowrap;
  }

  select {
    width: 100px;
    margin-left: 5px;
  }

  fieldset {
    border: 0;
    padding: 0;
    margin: 0;
    min-width: 0;
  }

  & > div {
    display: flex;

    &:last-of-type {
      margin-right: 10px;

      .bp3-button-group {
        margin-left: 5px;
      }
    }
  }

  .bp3-control-group {
    margin-right: 0.5rem;
  }

  .address-bar {
    width: 420px;
    transition: width 0.4s;

    input,
    span {
      -webkit-app-region: no-drag;
    }

    &.empty {
      width: 300px;
    }

    &:focus-within {
      span,
      input {
        border-color: @foreground-2;
      }
    }
  }

  & > .title {
    align-self: stretch;
    flex-shrink: 0;

    flex-flow: column nowrap;
    justify-content: center;
    user-select: none;

    margin: 0 10px;
  }
}
