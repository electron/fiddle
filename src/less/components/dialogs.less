@import "../variables.less";

.dialogs {
  z-index: 10;
  position: fixed;
  height: 100vh;
  width: 100vw;
  pointer-events: none;

  * {
    pointer-events: all;
  }
}

.dialog {
  background-color: @background-3;
  border: @border;
  color: @text-color-1;
  border-radius: 6px;
  padding: 10px;
  z-index: 100;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  animation: slide-up 0.4s ease;

  p {
    margin-bottom: 1.25rem;
    font-size: 1.25rem;
  }

  code {
    font-size: 12px;
    background: rgba(255, 255, 255, 0.1);
    padding: 2px;
    line-height: 12px;
  }

  &-buttons {
    margin-top: 1.25rem;
  }

  &.centered {
    position: absolute;
    width: 300px;
    left: calc(50% - (300px / 2));
    top: 100px;
  }

  &.arrow::after,
  &.arrow::before {
    border: solid transparent;
    content: " ";
    height: 0;
    width: 0;
    position: absolute;
    pointer-events: none;
  }

  &.arrow::after {
    border-color: rgba(136, 183, 213, 0);
    border-width: 9px;
  }

  &.arrow::before {
    border-color: rgba(194, 225, 245, 0);
    border-width: 10px;
  }

  &.arrow-top::after,
  &.arrow-top::before {
    bottom: 100%;
    left: 50%;
  }

  &.arrow-top::after {
    border-bottom-color: @background-3;
    margin-left: -9px;
  }

  &.arrow-top::before {
    border-bottom-color: @border-color-1;
    margin-left: -10px;
  }

  &.arrow-bottom::after,
  &.arrow-bottom::before {
    top: 100%;
    left: 50%;
  }

  &.arrow-bottom::after {
    border-top-color: @background-3;
    margin-left: -9px;
  }

  &.arrow-bottom::before {
    border-top-color: @border-color-1;
    margin-left: -10px;
  }

  &.arrow-left::after,
  &.arrow-left::before {
    right: 100%;
    top: 20px;
  }

  &.arrow-left::after {
    border-right-color: @background-3;
    margin-top: -9px;
  }

  &.arrow-left::before {
    border-right-color: @border-color-1;
    margin-top: -10px;
  }

  &.arrow-right::after,
  &.arrow-right::before {
    left: 100%;
    top: 20px;
  }

  &.arrow-right::after {
    border-left-color: @background-3;
    margin-top: -9px;
  }

  &.arrow-right::before {
    border-left-color: @border-color-1;
    margin-top: -10px;
  }

  button {
    color: #1e2527;
    float: right;
    margin-left: 9px;
    margin-bottom: 0;
  }

  a {
    cursor: pointer;
  }

  h4 {
    margin-bottom: 0;
  }

  p:last-of-type {
    margin-bottom: 0;
  }

  input {
    color: rgba(159, 234, 249, 0.5);
    height: 28px;
    margin-top: 10px;
    background-color: @background-1;
    border: 1px solid @border-color-2;
    outline: 0 solid transparent;

    &:focus {
      border-color: @foreground-2;
    }
  }
}

.dialogDrop {
  position: fixed;
  z-index: 100;
  top: 0;
  left: 0;
  width: 100vw;
  height: 100vh;
  background: black;
  transition: opacity 0.4s ease-in-out;
  pointer-events: all;
  opacity: 0.5;
}

.dialog-add-version {
  .bp3-file-input {
    width: 100%;
  }

  .bp3-callout {
    margin-top: 1rem;
  }
}

select {
  font-size: 12px;
  height: 28px;
  background-color: white;
  line-height: 15px;
}

@keyframes slide-up {
  0% {
    opacity: 0;
    transform: translateY(20px);
  }

  100% {
    opacity: 1;
    transform: translateY(0);
  }
}
