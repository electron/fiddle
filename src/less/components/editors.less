.mosaic-tile {
  animation: fadein 0.15s;
}

.mosaic-window-toolbar  {
  & > div {
    display: flex;
    justify-content: space-between;
    width: 100%;
    padding: 0 5px 0 5px;

    .mosaic-controls {
      transform: scale(0.75);
      transform-origin: right;
    }

    h5 {
      margin: 4px 0 0 0;
    }

    button {
      margin-left: 8px;
    }
  }
}

// TODO: support new file
// update if list of Editor ID changes
@editor-ids: main\.js, renderer\.js, index\.html, preload\.js, styles\.css;

// takes each editor ID and increase its z-index if the parent Mosaic root
// has focused__id class for that specific id.
each(@editor-ids, {
  .focused__@{value}.mosaic .mosaic-window.@{value} {
    z-index: 2;
  }
});

.mosaic-window {
  z-index: 1;
}

@keyframes fadein {
  from { opacity: 0; }
  to { opacity: 1; }
}
