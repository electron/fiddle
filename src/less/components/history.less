.revision-list {
  max-height: 400px;
  overflow-y: auto;
}

.revision-list ul {
  list-style-type: none;
  padding: 0;
  margin: 0;
}

.revision-item {
  padding: 10px;
  margin-bottom: 5px;
  border-radius: 3px;
  cursor: pointer;
  transition: background-color 0.2s ease;
  border-left: 3px solid #106ba3;
}

.revision-item:hover {
  background-color: rgba(167, 182, 194, 0.3);
}

.revision-content {
  display: flex;
  flex-direction: column;
}

.revision-icon {
  margin-right: 8px;
}

.sha-label {
  font-size: 12px;
  color: #738694;
  margin-left: 10px;
  font-family: monospace;
}

.revision-details {
  display: flex;
  justify-content: space-between;
  align-items: center;
  margin-top: 5px;
}

.revision-date {
  color: #738694;
  font-size: 12px;
}

.revision-changes {
  display: flex;
  gap: 5px;
}

.history-loading {
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 20px;
}
