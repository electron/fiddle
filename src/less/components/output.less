@import '../variables.less';

.output {
  font-size: 12px;
  font-family: @fonts-common;
  font-weight: 600;
  top: 10vh;
  padding: 0;
  height: 100%;
  -webkit-app-region: no-drag;

  .monaco-editor {
    color: var(--text-color-1);
    box-shadow: inset 0 0 100px 100px rgba(0, 0, 0, 0.15);
    border-top: @border;

    .margin,
    .monaco-editor-background {
      background: var(--background-2);
    }
  }
}
