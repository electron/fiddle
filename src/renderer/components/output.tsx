import * as React from 'react';

import { autorun, reaction } from 'mobx';
import { observer } from 'mobx-react';
import * as MonacoType from 'monaco-editor';
import {
  MosaicContext,
  MosaicNode,
  MosaicParent,
} from 'react-mosaic-component';

import { OutputEntry } from '../../interfaces';
import { AppState } from '../state';
import { WrapperEditorId } from './output-editors-wrapper';

interface CommandsProps {
  readonly appState: AppState;
  readonly monaco: typeof MonacoType;
  monacoOptions: MonacoType.editor.IEditorOptions;
  // Used to keep testing conform
  renderTimestamp?: (ts: number) => string;
}

/**
 * This component represents the "console" that is shown
 * whenever a Fiddle is launched in Electron.
 *
 * @class Output
 * @extends {React.Component<CommandsProps>}
 */
export const Output = observer(
  class Output extends React.Component<CommandsProps> {
    public static contextType = MosaicContext;
    public context: MosaicContext<WrapperEditorId>;
    public editor?: MonacoType.editor.IStandaloneCodeEditor;
    public language = 'consoleOutputLanguage';

    private outputRef = React.createRef<HTMLDivElement>();
    private readonly model: MonacoType.editor.ITextModel;

    // make it wide enough to fit all of OutputEntry's timestamps
    private readonly lineNumbersMinChars =
      new Date('2021-12-31T23:59:59').toLocaleTimeString().length + 1;

    constructor(props: CommandsProps) {
      super(props);

      const { monaco } = this.props;
      this.language = 'consoleOutputLanguage';
      this.model = monaco.editor.createModel('', this.language);
      this.updateModel();
      reaction(
        () => props.appState.output.length,
        () => this.updateModel(),
      );
    }

    public async componentDidMount() {
      autorun(async () => {
        await this.initMonaco();
        this.toggleConsole();
      });
    }

    public componentWillUnmount() {
      this.destroyMonacoEditor();
    }

    public componentDidUpdate() {
      this.toggleConsole();
    }

    /**
     * Initialize Monaco.
     */
    public async initMonaco() {
      const { monaco, monacoOptions: monacoOptions } = this.props;
      const ref = this.outputRef.current;
      if (ref) {
        this.setupCustomOutputEditorLanguage(monaco);
        this.editor = monaco.editor.create(ref, {
          language: this.language,
          theme: 'main',
          readOnly: true,
          contextmenu: false,
          automaticLayout: true,
          model: this.model,
          ...monacoOptions,
          wordWrap: 'on',
        });
      }
    }

    /**
     * Destroy Monaco.
     */
    public destroyMonacoEditor() {
      if (typeof this.editor !== 'undefined') {
        console.log('Editor: Disposing');
        this.editor.dispose();
        delete this.editor;
      }
    }

    public toggleConsole() {
      /**
       * Type guard to check whether a react-mosaic node is a parent in the tree
       * or a leaf. Leaf nodes are represented by the string value of their ID,
       * whereas parent nodes are objects containing information about the nested
       * binary tree.
       * @param node A react-mosaic node
       * @returns Whether that node is a MosaicParent or not
       */
      const isParentNode = (
        node: MosaicNode<WrapperEditorId> | null,
      ): node is MosaicParent<WrapperEditorId> => {
        return (node as MosaicParent<WrapperEditorId>)?.direction !== undefined;
      };

      const { isConsoleShowing } = this.props.appState;

      if (this.context.mosaicActions) {
        const mosaicTree = this.context.mosaicActions.getRoot();
        if (isParentNode(mosaicTree)) {
          // splitPercentage defines the percentage of space the first panel takes
          // e.g. 25 would mean the two children panels are split 25%/75%
          if (!isConsoleShowing) {
            mosaicTree.splitPercentage = 0;
          } else if (mosaicTree.splitPercentage === 0) {
            mosaicTree.splitPercentage = 25;
          }
          this.context.mosaicActions.replaceWith([], mosaicTree);
        }
      }
    }

    /**
     * Set up Monaco Language to catch custom regex expressions
     *
     * tokenizes timestamps so`main` theme can render them with a custom colour.
     */
    private setupCustomOutputEditorLanguage(monaco: typeof MonacoType) {
      // register a new language
      monaco.languages.register({ id: 'consoleOutputLanguage' });
      // Register a tokens provider for the language
      monaco.languages.setMonarchTokensProvider('consoleOutputLanguage', {
        tokenizer: {
          root: [[/\d{1,2}:\d{2}:\d{2}\s(A|P)M/, 'custom-date']],
        },
      });
    }

    /**
     * Sets the model and content on the editor
     *
     * @private
     * @memberof Output
     */
    private async updateModel() {
      // set the lines
      const lines = Output.getLines(this.props.appState.output);
      this.model.setValue(lines.map(({ text }) => text).join('\n'));

      // if we have an editor, tell it the line numbers and scroll to newest
      const { editor, lineNumbersMinChars } = this;
      if (!editor) return;
      const timestrs = lines.map(({ timeString }) => timeString);
      // adjust `i` here because the value passed in by monaco starts at 1, not 0
      const lineNumbers = (i: number) => timestrs[i - 1] || '';
      editor.updateOptions({ lineNumbers, lineNumbersMinChars });
      editor.revealLine(editor.getScrollHeight());
    }

    /**
     * An OutputEntry might span multiple lines.
     * Split it into individual lines to ensure each one has a timestamp.
     *
     * @param {OutputEntry} entries that may include paragraphs
     * @returns {OutputEntry} single-line entries
     * @memberof Output
     */
    private static getLines(paragraphs: OutputEntry[]): OutputEntry[] {
      const lines: OutputEntry[] = [];

      paragraphs = paragraphs.slice(-1000);

      for (const { text, timeString } of paragraphs) {
        for (const line of text.split(/\r?\n/)) {
          lines.push({ text: line, timeString });
        }
      }

      return lines;
    }

    public render(): JSX.Element | null {
      const { isConsoleShowing } = this.props.appState;

      return (
        <div
          className="output"
          ref={this.outputRef}
          style={{ display: isConsoleShowing ? 'inline-block' : 'none' }}
        />
      );
    }
  },
);
