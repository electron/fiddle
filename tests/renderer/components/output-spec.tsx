import * as React from 'react';

import { shallow } from 'enzyme';

import { Output } from '../../../src/renderer/components/output';
import { StateMock } from '../../mocks/state';

const mockContext = {
  mosaicActions: {
    expand: jest.fn(),
    remove: jest.fn(),
    hide: jest.fn(),
    replaceWith: jest.fn(),
    updateTree: jest.fn(),
    getRoot: jest.fn(),
  },
  mosaicId: 'output',
};

describe('Output component', () => {
  let store: any;
  let monaco: any;

  beforeEach(() => {
    store = new StateMock();
    ({ monaco } = (window as any).ElectronFiddle);
  });

  it('renders the output container', () => {
    const wrapper = shallow(
      <Output appState={store} monaco={monaco} monacoOptions={{}} />,
    );
    expect(wrapper.html()).toBe(
      '<div class="output" style="display:inline-block"></div>',
    );
  });

  it('correctly sets the language', () => {
    const wrapper = shallow(
      <Output appState={store as any} monaco={monaco} monacoOptions={{}} />,
    );

    expect((wrapper.instance() as any).language).toBe('consoleOutputLanguage');
  });

  describe('initMonaco()', () => {
    it('attempts to create an editor', async () => {
      const wrapper = shallow(
        <Output appState={store as any} monaco={monaco} monacoOptions={{}} />,
      );
      const instance: any = wrapper.instance();

      instance.outputRef.current = 'ref';
      await instance.initMonaco();

      expect(monaco.editor.create).toHaveBeenCalled();
      expect(monaco.editor.createModel).toHaveBeenCalled();
    });
  });

  it('componentWillUnmount() attempts to dispose the editor', async () => {
    const wrapper = shallow(
      <Output appState={store as any} monaco={monaco} monacoOptions={{}} />,
    );
    const instance: any = wrapper.instance();

    instance.outputRef.current = 'ref';
    await instance.initMonaco();
    instance.componentWillUnmount();

    expect(monaco.latestEditor.dispose).toHaveBeenCalled();
  });

  it('hides the console with react-mosaic-component', async () => {
    // manually trigger lifecycle methods so that
    // context can be set before mounting method
    const wrapper = shallow(
      <Output appState={store} monaco={monaco} monacoOptions={{}} />,
      {
        context: mockContext,
        disableLifecycleMethods: true,
      },
    );
    const instance: any = wrapper.instance();

    // Todo: There's a scary bug here in Jest / Enzyme. At this point in time,
    // the context is {}. That's never the case in production.
    // direction is required to be recognized as a valid root node
    mockContext.mosaicActions.getRoot.mockReturnValue({
      splitPercentage: 25,
      direction: 'row',
    });

    wrapper.instance().context = mockContext;
    wrapper.instance().componentDidMount!();

    instance.outputRef.current = 'ref';
    await instance.initMonaco();

    expect(mockContext.mosaicActions.replaceWith).toHaveBeenCalled();
    expect(mockContext.mosaicActions.replaceWith).toHaveBeenCalledWith(
      [],
      expect.objectContaining({ splitPercentage: 25 }),
    );
    expect(wrapper.html()).not.toBe(null);
  });

  it('updateModel updates model with correct values', async () => {
    store.output = [
      {
        text: 'Hi!',
        timeString: '12:00:01 PM',
      },
      {
        isNotPre: true,
        text: 'Hi!',
        timeString: '12:00:01 PM',
      },
    ];

    const wrapper = shallow(
      <Output appState={store as any} monaco={monaco} monacoOptions={{}} />,
    );
    const instance: any = wrapper.instance();

    instance.outputRef.current = 'ref';
    await instance.initMonaco();
    instance.updateModel();

    expect(monaco.editor.createModel).toHaveBeenCalled();
    expect(instance.editor.revealLine).toHaveBeenCalled();
  });

  it('updateModel correctly observes and gets called when output is updated', async () => {
    store.output = [
      {
        timestamp: 1532704073130,
        text: 'Hi!',
      },
    ];

    const wrapper = shallow(
      <Output appState={store as any} monaco={monaco} monacoOptions={{}} />,
    );

    const instance: any = wrapper.instance();
    const spy = jest.spyOn(instance, 'updateModel');

    instance.outputRef.current = 'ref';
    await instance.initMonaco();

    instance.updateModel();

    // new output
    store.output = [
      {
        timestamp: 1532704073130,
        text: 'Hi!',
      },
      {
        timestamp: 1532704073130,
        text: 'Hi!',
        isNotPre: true,
      },
    ];
    expect(spy).toHaveBeenCalledTimes(2);
  });

  it('handles componentDidUpdate', async () => {
    // set up component
    const wrapper = shallow(
      <Output appState={store as any} monaco={monaco} monacoOptions={{}} />,
    );
    const instance: any = wrapper.instance();
    const spy = jest.spyOn(instance, 'toggleConsole');

    instance.outputRef.current = 'ref';
    await instance.initMonaco();

    // setContent will trigger componentDidUpdate()
    instance.editor.setContent(store.output);
    expect(spy).toHaveBeenCalled();
  });
});
