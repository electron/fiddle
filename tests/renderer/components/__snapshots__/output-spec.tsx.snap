// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Output component renders with output 1`] = `
<div
  className="output"
>
  <p
    key="1532704073130--0--0"
    style={Object {}}
  >
    <span
      className="timestamp"
    >
      1532704073130
    </span>
    Hi!
  </p>
  <p
    key="1532704073130--1--0"
    style={
      Object {
        "whiteSpace": "initial",
      }
    }
  >
    <span
      className="timestamp"
    >
      1532704073130
    </span>
    Hi!
  </p>
</div>
`;
